feat: reading multiple pdf files with a single PDFParser object

nicolabaesso opened the pull request:
Elements changed:

1. Added new test case in a separate file
2. Added the two example PDFs
3. Add the reset of the pages array when the data variable is null

I've added this elements because in my corporate job we are using this library, and recreating everytime the PDFParser object is not something I'm a fan of.
Other test cases are not failing, so no regressions.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a new command-line option to enable singleton mode for the PDF parser, allowing reuse of a single parser instance.
  * Introduced new methods to reset the internal state of the PDF parser without recreating the entire object.

* **Bug Fixes**
  * Improved handling of multiple PDF files parsed sequentially to ensure correct data extraction.

* **Tests**
  * Added tests to verify parsing multiple PDFs with the same structure using a single parser instance.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

modesty commented on pdfparser.js:
```diff
@@ -117,7 +117,13 @@ export default class PDFParser extends EventEmitter {
 			this.emit("pdfParser_dataReady", this.#data);
 		} else {
-			this.#data = { ...this.#data, ...data };
+			if(this.#data===data){
```
when could this occur? this.#data is a collection of data from all data events.

nicolabaesso replied:
This was a mistake left in the commit, my bad. Removed

modesty reviewed (commented):
thanks for adding more tests. A few thoughts on making the instance of PDFJSClass reusable:
1. pdfparser instance (or the client that instantiates PDFJSClass) needs to be reset/reusable whenever PDFParser is created. (line 107 of pdfparser.js)
2. lib/pdf.js: setting this.pages=[] is not sufficient to dispose the object, pdfDocument and rawTextContents needs reset too. Recomment to call existing `destroy` method.

nicolabaesso commented:
Hi @modesty,
thank you for your review. As suggested, I removed the this.pages=[] line and instead called the already available destroy() function.
Also I've added a function to reset the PDFJS object, is this what you were mentioning? Otherwise let me know.
I've also removed the if in line 120 of pdfparser.js, it was a leftover of one test I was doing to understand the code.

nicolabaesso commented:
Hi @modesty,
sorry for the pressure, can you give me a feedback on the code? The next days I could make some changes if something is still wrong (I feel like the function for resetting the PDFJS object could use some more work, but I would like to have your opinion).

Thank you!

modesty commented:
sorry for the delay. code LGTM.
could you merge master into this branch and run all tests when ready ?

nicolabaesso commented:
> sorry for the delay. code LGTM. two notes:
> 
>     1. could you add a command line option to enable it optionally? default `false` would keep current clients intact.
> 
>     2. could you merge master into this branch and run all tests when ready?

1. Sorry, I'm not getting it: how we can provide this feature when using the library via command line? Should it be a whole section where, if the flag is provided, you can pass two or more files and the library reads them with a single instance? Just asking

2. Master branch merged, I modified the reset part by calling a similar method to the destroy() one, with the exception of not removing the listeners. Keeping the same method hanged the test forever, with this change the test cases are ok and done in less than 5 seconds.

Let me know what to do next, I'll do it as soon as I can.

Thank you!

nicolabaesso commented:
Merged recent changes of master, tests are ok

nicolabaesso commented:
Added a little change to package.json: issue #381 says that the semver is broken in 3.1.5.
Test are still working fine, it's better to make a major release as suggested on the issue.
I can set up some scripts to automate the semver processing, just let me know if it's a modification you could appreciate.

Thank you!

nicolabaesso commented:
Hi @modesty,
it's been a while since your last comment, do you have any feedback for this PR?

Thank you

modesty commented on package.json:
```diff
@@ -55,13 +55,13 @@
+	  "devEngines": {
+		  "packageManager": {
+			  "name": "npm",
+			  "version": ">=8.19.2"
```
no downgrade of node & npm please

nicolabaesso replied:
Restored to your last changes. BTW comment #381 underlines that you're introducing a breaking change, maybe this could be version 4.0.0?
Minor and patches release should not introduce breaking changes

modesty commented on src/cli/p2jcli.ts:
```diff
@@ -123,10 +124,14 @@ class PDFProcessor {
-			this.pdfParser.on("pdfParser_dataError", (evtData: any) =>
-				this.onPrimaryError(evtData.parserError, reject)
-			);
+			if(!SINGLETON_PDF_PARSER || !this.pdfParser){
```
in a batch process, parseOnePDFStream will be invoked multiple times. Line 127 makes this.pdfParser always a singleton regardless of SINGLETON_PDF_PARSER. No change of current behavior if SINGLETON_PDF_PARSER not set (default) please, also check other places that instantiates pdfParser instances.

nicolabaesso replied:
My bad, modified the part.
Now it does the following:

- If the parameter is provided AND the pdfParser object is not initialized, the pdfParserObject gets initialized
- If the parameter is NOT provided, the pdfParser object gets initialized
- If the parameter is provided AND the pdfParser object is initialized, the pdfParserObject DOES NOT get initialized

Sorry for the previous mistake, is this the desired behaviour?

modesty commented on src/cli/p2jcli.ts:
```diff
@@ -145,10 +150,14 @@ class PDFProcessor {
-			this.pdfParser.on("pdfParser_dataError", (evtData: any) => {
-				this.onPrimaryError(evtData.parserError, reject);
-			});
+			if(!SINGLETON_PDF_PARSER || !this.pdfParser){
```
in a batch process, parseOnePDF will be invoked multiple times. Line 153 makes this.pdfParser always a singleton regardless of SINGLETON_PDF_PARSER. No change of current behavior if SINGLETON_PDF_PARSER not set (default) please, also check other places that instantiates pdfParser instances.

nicolabaesso replied:
Refer to my comment on parseOnePDFStream method (the same thing got applied, maybe we could use a refactor?)

modesty reviewed (changes requested):
there are unresolved comments.
